### PR TITLE
Making typed.List a typing Generic

### PR DESCRIPTION
**Pull request by luk-f-a**
This PR adds typing information to the typed List, so in can be used as a Generic in variable annotations, e.g.

```python
from numba.typed import List

my_list:List[int] = List()
my_list.append(1)
my_list.append('a') # would be rejected by mypy

for i in my_list:
    a: str = i # would be rejected by mypy
    print(a)
```

While some of these would also be runtime errors, this PR allows List to be fully checked as a generic by mypy.

This is not related to checking Numba itself with mypy, this just allows users to check their typed.List code with mypy.

Besides setting List as a Generic, it also annotates the main methods (append, getitem, setitem), and the methods for protocols: Iterable[T], Sized, Container[T], and Collection[T].

### CODE REVIEW

**Comment by esc** [comment]
@luk-f-a thanks for reporting this, I have added it to the review queue for now.

**Comment by luk-f-a** [comment]
@EPronovost would you have time to review this PR?

File: numba/typed/typedlist.py
Comment (EPronovost) [inline numba/typed/typedlist.py:324]: To be fully annotated these should have a `-> None` as well 
https://github.com/python/typeshed/blob/master/stdlib/3/builtins.pyi#L694-L738

File: numba/typed/typedlist.py
Comment (EPronovost) [inline numba/typed/typedlist.py:358]: Type annotation for `i`?

File: numba/typed/typedlist.py
Comment (EPronovost) [inline numba/typed/typedlist.py:384]: Type annotations for `start` and `stop`?

File: numba/typed/typedlist.py
Comment (EPronovost) [inline numba/typed/typedlist.py:29]: To stick with style elsewhere should we `import typing as pt` instead?

**Review by EPronovost — CHANGES_REQUESTED** [review]

File: numba/typed/typedlist.py
Comment (luk-f-a) [reply numba/typed/typedlist.py:324]: you're right!

File: numba/typed/typedlist.py
Comment (luk-f-a) [reply numba/typed/typedlist.py:358]: fixed

File: numba/typed/typedlist.py
Comment (luk-f-a) [reply numba/typed/typedlist.py:384]: fixed

File: numba/typed/typedlist.py
Comment (luk-f-a) [reply numba/typed/typedlist.py:29]: yes! I created this PR before we decided that, and forgot about it. thanks!

**Comment by luk-f-a** [comment]
@EPronovost thanks for the review. I fixed what you found, plus other similar things, eg other methods returning None which I hadn't annotated. I also completed those annotations that were incomplete, those where I had only annotated one input variable, like getitem, insert, index.

**Comment by luk-f-a** [comment]
Interesting finding from typeguard after I annotated `typed.List`

```
======================================================================
ERROR: test_extend_empty_unrefined (numba.tests.test_typedlist.TestExtend)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/vsts/work/1/s/numba/tests/test_typedlist.py", line 779, in test_extend_empty_unrefined
    l.extend(tuple())
  File "/home/vsts/miniconda3/envs/travisci/lib/python3.8/site-packages/typeguard/__init__.py", line 841, in wrapper
    check_return_type(retval, memo)
  File "/home/vsts/miniconda3/envs/travisci/lib/python3.8/site-packages/typeguard/__init__.py", line 665, in check_return_type
    raise exc from None
  File "/home/vsts/miniconda3/envs/travisci/lib/python3.8/site-packages/typeguard/__init__.py", line 663, in check_return_type
    check_type('the return value', retval, memo.type_hints['return'], memo)
  File "/home/vsts/miniconda3/envs/travisci/lib/python3.8/site-packages/typeguard/__init__.py", line 620, in check_type
    raise TypeError(
TypeError: type of the return value must be NoneType; got numba.typed.typedlist.List instead
```

Reproducer (it should always return None but in some cases it's returning self)

```python
from numba.typed import List
  ...: my_list = List([1,2])
  ...: print(my_list.extend(List([1,2])))
  ...: print(my_list.extend(tuple()))
None
[1, 2, 1, 2]
```

File: numba/typed/typedlist.py
Comment (EPronovost) [inline numba/typed/typedlist.py:392]: Under strict mode inferred optionals are disabled, so these should be `Optional[int]`

File: numba/typed/typedlist.py
Comment (EPronovost) [inline numba/typed/typedlist.py:339]: Should this be `List[T]` as it is here? https://github.com/python/typeshed/blob/master/stdlib/3/builtins.pyi#L720

**Review by EPronovost — CHANGES_REQUESTED** [review]
I wonder if we can/should add test coverage for this beyond just using `typeguard` checking.  For a full test of annotations, I think you need both positive and negative examples.  I've recently done similar work for `typeguard` [here](https://github.com/agronholm/typeguard/tree/master/tests/mypy).  To keep it smaller scope we could also just have positive tests; i.e. a test file that uses `List` with annotations and gets analyzed by mypy as part of CI.
I'm curious to hear what the core devs think about this additional testing for type annotations intended to be used by users. cc: @sklam @esc @stuartarchibald 
Note that if we want mypy to consider these type annotations in users' projects we need to add a `py.typed` file https://www.python.org/dev/peps/pep-0561/#packaging-type-information

File: numba/typed/typedlist.py
Comment (luk-f-a) [reply numba/typed/typedlist.py:392]: thanks, I didn't realize that strict optionals was the default, I thought that inferred optionals were the default.

Comment (luk-f-a) [reply numba/typed/typedlist.py:392]: I wonder if we shouldn't consider switching to `--no-strict-optional` to avoid the added verbosity. For example, `index` could fit in one line (easier to read) without losing much in terms of typing (I think `:int = None` is clear enough). what do you think?

```python
    def index(self, item: T, start: pt.Optional[int] = None,
              stop: pt.Optional[int] = None) -> int:
```
vs
```python
def index(self, item: T, start: int = None, stop: int = None) -> int:
```

File: numba/typed/typedlist.py
Comment (luk-f-a) [reply numba/typed/typedlist.py]: yes, somehow I hadn't thought of it when writing it as text, but it makes sense.

**Comment by luk-f-a** [comment]
@EPronovost I made some fixes following your latest comments, and also improved other things:
- I had only partially annotated `__setitem__`, now it has full annotations with overloads.
- in order to improve legibility by keeping signatures to one line, I created two type aliases.
- I ran mypy on this file and added a few fixes to remove fails
    - added  `# type: ignore[override]`. Without this mypy complains that the method conflicts with the supertype MutableSequence (a problem with the contravariance required by Liskov's substitution principle).
    - in the annotation to `extend` I had to change the input from `Iterable` to `Sequence` because the method calls `len(iterable)` and `iterable[0]`. 

Related to this last point (running mypy), the file now passes basic tests and it could included in "level 3" files. However, I didn't actually change `mypy.ini` because I'd like to do it on your improved version of `mypy.ini` from #6222. What's the best way to do it? Wait for it to be merged or just copy and paste your version here?

File: numba/typed/typedlist.py
Comment (EPronovost) [reply numba/typed/typedlist.py:401]: For internal annotations that's something we can decide on.  I think for external facing annotations we should be as precise as possible to avoid incompatibilities for users who don't use `--no-strict-optional`.

**Comment by luk-f-a** [comment]
> Note that if we want mypy to consider these type annotations in users' projects we need to add a py.typed file https://www.python.org/dev/peps/pep-0561/#packaging-type-information

I don't really understand this PEP. Does it mean that without a `py.typed` file someone with a project that imports numba would not be able to make use of the `typed.List`annotation when running mypy on their project?

**Comment by EPronovost** [comment]
Sorry for the delay. Yes to your above question.  There's relevant code in the mypy module finder here: https://github.com/python/mypy/blob/master/mypy/modulefinder.py#L180-L194

**Comment by luk-f-a** [comment]
> Sorry for the delay. Yes to your above question. There's relevant code in the mypy module finder here: https://github.com/python/mypy/blob/master/mypy/modulefinder.py#L180-L194

thanks. I've added the `py.typed` file in the latest commit. I'm still thinking how I could test that it's working correctly. Before I added `py.typed` I ran the following

```python
from numba.typed import List


l:List[int] = List([1, 2])
a=l[0]
l.append(2)
reveal_locals()
reveal_type(l)
```
and the result was correct (even without `py.typed`). So I don't know how to check whether `py.typed` is making a difference now.

File: numba/typed/typedlist.py
Comment (luk-f-a) [reply numba/typed/typedlist.py]: you're right, I changed it to the explicit `Optional`.

File: numba/typed/typedlist.py
Comment (sklam) [inline numba/typed/typedlist.py:364]:
```suggestion
    def __delitem__(self, i: Int_or_Slice) -> None:
```

File: numba/typed/typedlist.py
Comment (sklam) [inline numba/typed/typedlist.py:375]:
```suggestion
    def pop(self, i: "pt.SupportsIndex" = -1) -> T:
```

**Review by sklam — COMMENTED** [review]
`typeguard` is failing for `numba.tests.test_typedlist`. On CI, the test slices didn't run any of the failing tests by "luck".

File: numba/typed/typedlist.py
Comment (luk-f-a) [reply numba/typed/typedlist.py]: Should the same be used in getitem and setitem? From your explanation it sounds like it should

**Comment by sklam** [comment]
Test is failing because `Protocol` is also new on py3.8.

**Comment by luk-f-a** [comment]
> `typeguard` is failing for `numba.tests.test_typedlist`. On CI, the test slices didn't run any of the failing tests by "luck".

what's the solution? Did you run the full suite locally? Is it possible to expand typeguard to run on more slices?

**Comment by luk-f-a** [comment]
> Test is failing because `Protocol` is also new on py3.8.

I tried a fix but made it worse. I'll build something better tomorrow.

**Comment by sklam** [comment]
> what's the solution? Did you run the full suite locally? Is it possible to expand typeguard to run on more slices?

Right, we will need to make more configs use typeguard.

**Comment by sklam** [comment]
If we guard the protocol type for py3.8 only and use a string to refer to it, we can avoid changing the implementation of extend.

something like
```python
if python_version >= (3, 8):
    class _Sequence(pt.Protocol[T_co]): ...

class List:
    def extend(self, iterable: "_Sequence"): ...
```

**Comment by sklam** [comment]
I think it's ok to do type-checks on only py>=3.8. I'm seeing that `Protocol` is essential.

**Comment by luk-f-a** [comment]
Thanks @sklam, I'll try what you suggested above. Funny you found a 5th option, I was considering the pros/cons of:
- keeping extend untyped until 3.8 is the minimum supported version
- changing extend such that it does not require len and getitem (I tried this yesterday and failed) so it can be types as Iterable
- requiring typing_extensions as a dependency (cons: not worth adding a dependency for this)
- typing with _Sequence but not with _Sequence[T], and _Sequence is Any for <3.8

**Review by sklam — APPROVED** [review]

**Pull request merged**
